runtime/service.go

```diff
@@ -1180,44 +1181,63 @@ func (s *service) shutdown(
 	timeout time.Duration,
 	req *taskAPI.ShutdownRequest,
 ) error {
-	shutdownCtx, cancel := context.WithTimeout(requestCtx, timeout)
-	defer cancel()
+	s.logger.Info("stopping the VM")
+
 	go func() {
-		// Once the shutdown procedure is done, the shim needs to shutdown too.
-		// This also ensures that if the VMM is still alive, it will receive a
-		// SIGKILL via exec.CommandContext
-		<-shutdownCtx.Done()
+		s.shutdownLoop(requestCtx, timeout, req)
 		s.shimCancel()
 	}()
 
-	s.logger.Info("stopping the VM")
-
-	// Try to tell agent to exit, causing the VM to begin a reboot. If that
-	// fails, try to forcibly stop the VMM. If that too fails, just cancel
-	// the shutdownCtx to fast-path to the VMM getting SIGKILL.
-	_, shutdownErr := s.agentClient.Shutdown(shutdownCtx, req)
-	if shutdownErr != nil {
-		s.logger.WithError(shutdownErr).Error("failed to shutdown VM agent")
-		stopVMMErr := s.machine.StopVMM()
-		if stopVMMErr != nil {
-			s.logger.WithError(stopVMMErr).Error("failed to forcibly stop VMM")
-			cancel()
-		}
-	}
-
-	// wait for the shimCtx to be done, which means the VM has exited and we're ready
-	// to shutdown
-	<-s.shimCtx.Done()
-	if shutdownCtx.Err() == context.DeadlineExceeded {
-		return status.Error(codes.DeadlineExceeded,
-			"timed out waiting for VM shutdown, VMM was sent SIGKILL")
+	err := s.machine.Wait(context.Background())
+	if err == nil {
+		return nil
 	}
+	return status.Error(codes.DeadlineExceeded, fmt.Sprintf("the VMM was killed forcibly: %v", err))
+}
 
-	if s.vmExitErr != nil {
-		return status.Error(codes.Internal,
-			fmt.Sprintf("VMM exit errors: %v", s.vmExitErr))
+func (s *service) shutdownLoop(
+	requestCtx context.Context,
+	timeout time.Duration,
+	req *taskAPI.ShutdownRequest,
+) error {
+	actions := []struct {
+		name     string
+		shutdown func() error
+		timeout  time.Duration
+	}{
+		{
+			name: "send a request to the in-VM agent",
+			shutdown: func() error {
+				_, err := s.agentClient.Shutdown(requestCtx, req)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			timeout: timeout,
+		},
+		{
+			name: "stop the jailer",
+			shutdown: func() error {
+				return s.jailer.Stop()
+			},
+			timeout: defaultSigtermTimeout,
+		},
+	}
+
+	for _, action := range actions {
+		s.logger.Debug(action.name)
+		err := action.shutdown()
+		if err != nil {
+			s.logger.WithError(err).Errorf("failed to %s", action.name)
+		} else if action.timeout != 0 {
```

Comment:
Can this `else if` be just an `if` statement separate from the one above. They aren't really related.

Reply:
It is related. If sending a request is not succeeding, I'd like to skip `time.Sleep()`

Reply:
I think we could get rid of the if part and just let time.Sleep() be called since it's safe to call with a 0 duration.

Reply:
I thought Go would do something but you are right. Will remove the if.

https://golang.org/src/runtime/time.go?h=timeSleep#L84

Reply:
Done.